### ISSUE TITLE
oracle fails with "ORA-03146: invalid buffer length for TTC field" for mergeInto with larger CLOB

### ISSUE THREAD

**Issue by torstenwerner** [jOOQ/jOOQ#13840]
### Expected behavior

Not sure. 😉 The best solution is an automatic workaround provided by jooq. But it looks like a bug in the oracle database. In case an automatic workaround is not possible I suggest to add some documentation about the problem and how to implement a workaround manually.

### Actual behavior

Please check https://github.com/torstenwerner/jOOQ-mcve – the java version.

The test method _failsWithOra03146ForSomeLargerData_ uses mergeInto to insert/update a row but it fails if the string for the clob gets too large. The error is "ORA-03146: invalid buffer length for TTC field".

The test method _workaroundThatWorksForReallyLargeData_ implements a workaround that inserts/updates the row without the clob and it executes an update to add the clob afterwards. It works for really huge data.

### Steps to reproduce the problem

- Execute the test class JavaTest.

### Versions

- jOOQ: 3.16.5
- Java: 11
- Database (include vendor): oracle 21 xe
- OS: macos
- JDBC Driver (include name if unofficial driver): com.oracle.database.jdbc:ojdbc11:21.6.0.0.1

**Comment by lukaseder** [comment]
Thanks a lot for your detailed report and the MCVE. I'll investigate this more thoroughly, soon. I think that jOOQ should be able to work around this as soon as the problem is understood. You might be able to implement a workaround already now by implementing a data type binding that replaces the default behaviour on your `CLOB` column:
https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types/codegen-database-forced-types-binding/

**Comment by lukaseder** [comment]
Thanks for your patience. I'm looking into this now. I can reproduce the problem with jOOQ 3.18.0-SNAPSHOT

**Comment by lukaseder** [comment]
This works as expected:

```sql
declare
  c clob;
begin 
  for i in 1 .. 1024 loop
    c := c || '01234567';
  end loop;

  merge into mcve 
  using (select 1 from dual) 
  on (id = 42)
  when matched then update set data = c
  when not matched then insert (id, data) values (42, c);
end;
```

**Comment by lukaseder** [comment]
The problem can also be reproduced with JDBC directly like this:

```java
try (PreparedStatement s = connection.prepareStatement("""
      merge into mcve.mcve
      using (select 1 from dual)
      on (id = 42)
      when matched then update set data = ?
      when not matched then insert (id, data) values (42, ?)

""")) {
    s.setString(1, "01234567".repeat(1024));
    s.setString(2, "01234567".repeat(1024));
    s.executeUpdate();
}
```

**Comment by lukaseder** [comment]
This works:

```java
try (PreparedStatement s = connection.prepareStatement("""
      merge into mcve.mcve
      using (select 1 from dual)
      on (id = 42)
      when matched then update set data = ?
      when not matched then insert (id, data) values (42, ?)

""")) {
    Clob c = connection.createClob();
    c.setString(1, "01234567".repeat(1024));
    try {
        s.setClob(1, c);
        s.setClob(2, c);
        s.executeUpdate();
    }
    finally {
        c.free();
    }
}
```

It seems the built-in `org.jooq.impl.ClobBinding` could help here? Have you tried using that as a `<forcedType/>`? I'll try looking into this a bit more in depth next week. Perhaps, we should auto-bind such values as `Clob` rather than `String`. Though, it does seem like an ojdbc bug...

**Comment by torstenwerner** [comment]
No I haven't tried anything with `<forcedType/>` here.

**Comment by lukaseder** [comment]
Related: https://github.com/jOOQ/jOOQ/issues/6516

**Comment by lukaseder** [comment]
It seems the built-in `ClobBinding` won't solve the problem here, because it has a flaw: https://github.com/jOOQ/jOOQ/issues/14027, which results in it delegating to the `DefaultStringBinding` again, which binds the `CLOB` as a `String`.

This isn't a problem in almost every statement, but does seem to cause issues with `MERGE`, probably due to some ojdbc bug.

So, workarounds include:

- Wait for the fix of https://github.com/jOOQ/jOOQ/issues/14027, and use `ClobBinding`
- Implement your own `ClobBinding` as in the fix of https://github.com/jOOQ/jOOQ/issues/14027

As far as this issue here is concerned, we can probably patch the effective data type binding for use with Oracle's `MERGE` statements. I'm not sure yet how bold this "patch" should be. In principle, most JDBC `ps.setClob()` calls can be replaced by `ps.setString()` calls, which is what jOOQ has done for almost 10 years now, successfully. Binding `CLOB` explicitly with ojdbc is a thing of the past, but as this issue shows, there seem to be bugs, still.

Reverting the calls to `ps.setClob()` globally seems risky, there might be other side effects / regressions, so I wouldn't want to "fix" this by binding `CLOB` everywhere. I'll think about this a bit more...

**Comment by lukaseder** [comment]
Possibly related:

- https://github.com/jOOQ/jOOQ/issues/15595

**Comment by lukaseder** [comment]
This seems to have been fixed indeed with https://github.com/jOOQ/jOOQ/issues/15595. The issue no longer reproduces with 3.16.23-SNAPSHOT.

Fixed in versions, 3.19.0, 3.18.8, 3.17.17, 3.16.23

Closing this as a duplicate of #15595
